File: 304.range-sum-query-2-d-immutable.cpp
/*
 * @lc app=leetcode id=304 lang=cpp
 *
 * [304] Range Sum Query 2D - Immutable
 */
#include <vector>

using std::vector;

// @lc code=start
class NumMatrix {
    vector<vector<int>> psum;  // S: O(m*n)

public:
    // T: O(m*n)
    NumMatrix(vector<vector<int>>& matrix) {
        const int m = matrix.size(), n = matrix[0].size();
        psum = vector<vector<int>>(m + 1, vector<int>(n + 1, 0));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                psum[i + 1][j + 1] = psum[i + 1][j] + psum[i][j + 1] - psum[i][j] + matrix[i][j];
            }
        }
    }
    
    // T: O(1)
    int sumRegion(int row1, int col1, int row2, int col2) {
        return psum[row2 + 1][col2 + 1] - psum[row2 + 1][col1] - psum[row1][col2 + 1] + psum[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */
// @lc code=end

File: 304.range-sum-query-2-d-immutable.py
#
# @lc app=leetcode id=304 lang=python3
#
# [304] Range Sum Query 2D - Immutable
#
from typing import *

# @lc code=start
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        m = len(matrix)
        n = len(matrix[0])
        self.psum = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m):
            for j in range(n):
                self.psum[i + 1][j + 1] = self.psum[i + 1][j] + self.psum[i][j + 1] - self.psum[i][j] + matrix[i][j]

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return self.psum[row2 + 1][col2 + 1] - self.psum[row2 + 1][col1] - self.psum[row1][col2 + 1] + self.psum[row1][col1]


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)
# @lc code=end

